torchci/lib/jobUtils.ts

```diff
@@ -144,3 +151,70 @@ export function isSameFailure(
     isEqual(jobA.failure_captures, jobB.failure_captures)
   );
 }
+
+export async function removeCancelledJobAfterRetry<T extends BasicJobData>(
+  jobs: T[]
+): Promise<T[]> {
+  // When a worlflow is manually cancelled and retried, the leftover cancel signals from
+  // the previous workflow run are poluting HUD and Dr.CI. For example, the pull request
+  // https://hud.pytorch.org/pytorch/pytorch/pull/107339 had many cancelled binary build
+  // jobs showing up as new failures after the workflow had been retried successfully.
+  //
+  // The issue here is that the cancelled job name is not the same as the successfully
+  // retried one, for example manywheel-py3_10-cuda11_8-test (cancel) was retried as
+  // manywheel-py3_10-cuda11_8-test / test (success). As their names look different,
+  // HUD and Dr.CI treat them incorrectly as two different jobs and mark the cancelled
+  // one as a failure.
+  //
+  // So the fix here is to check if a cancelled job has been retried successfully and
+  // keep or remove it from the list accordingly.
+  const trie: TrieSearch<T> = new TrieSearch<T>("name", {
+    splitOnRegEx: /\s\/\s/g,
```

Comment:
Does this work correctly with sharded jobs? It seems like it might use a successful run of shard 1 to overwrite a failed run on shard 2...

Reply:
This should work with shard job because the full name of the job is used.  You could see that the name from Rockset is used as it is without any truncating.  So doing a trie search for `pull / linux-focal-cuda12.1-py3.10-gcc9 / test (default, 1, 5, linux.4xlarge.nvidia.gpu)` in `WORKFLOW NAME / JOB NAME` format will never match with other shards because the shard id `1` is there.

On the other hand, job name like `linux-focal-py3.8-clang10` is actually wrong and only show up when the job is cancelled prematurely.  The correct name should have a suffix like `linux-focal-py3.8-clang10 / build` or `C linux-focal-py3.8-clang10 / test (...)`

In the example, https://hud.pytorch.org/pr/105055 has a bunch of cancel jobs.  If you search for `linux-focal-py3.8-clang10` after the change https://torchci-git-fork-huydhn-remove-wrong-cancel-948947-fbopensource.vercel.app/pr/105055, you can see all the shards are there.

Reply:
I have added a test case to cover this case